src/Cli/dotnet/commands/dotnet-workload/update/WorkloadUpdateCommand.cs

```diff
@@ -143,26 +239,33 @@ public void UpdateWorkloads(bool includePreviews = false, DirectoryPath? offline
             transaction.Run(
                 action: context =>
                 {
+                    UpdateInstalledWorkloadsFromHistory(sdkFeatureBand, context, offlineCache);
+
                     bool rollback = !string.IsNullOrWhiteSpace(_fromRollbackDefinition);
 
                     foreach (var manifestUpdate in manifestsToUpdate)
                     {
-                        _workloadInstaller.InstallWorkloadManifest(manifestUpdate, context, offlineCache, rollback);
+                        if (manifestUpdate.NewFeatureBand != null && manifestUpdate.NewVersion != null &&
+                            (manifestUpdate.ExistingFeatureBand is null ||
+                            !manifestUpdate.ExistingVersion.Equals(manifestUpdate.NewVersion) ||
+                            !manifestUpdate.ExistingFeatureBand.ToString().Equals(manifestUpdate.NewFeatureBand)))
+                        {
+                            _workloadInstaller.InstallWorkloadManifest(manifestUpdate, context, offlineCache, rollback);
+                        }
                     }
 
                     _workloadResolver.RefreshWorkloadManifests();
 
-                    var workloads = GetUpdatableWorkloads();
-
-                    _workloadInstaller.InstallWorkloads(workloads, sdkFeatureBand, context, offlineCache);
-
-                    if (useRollback)
+                    if (string.IsNullOrWhiteSpace(_fromHistorySpecified) || !_historyManifestOnlyOption)
                     {
-                        _workloadInstaller.WriteInstallState(sdkFeatureBand, GetInstallStateContents(manifestsToUpdate));
+                        var workloads = GetUpdatableWorkloads();
+                        _workloadInstaller.InstallWorkloads(workloads, sdkFeatureBand, context, offlineCache);
                     }
-                    else
+
+                    UpdateInstallState(shouldUpdateInstallState, manifestsToUpdate, _workloadInstaller.GetFailingWorkload());
+                    if (!string.IsNullOrWhiteSpace(_fromHistorySpecified))
                     {
-                        _workloadInstaller.DeleteInstallState(sdkFeatureBand);
+                        _workloadInstaller.GarbageCollect(workloadSetVersion => _workloadResolverFactory.CreateForWorkloadSet(_dotnetPath, _sdkVersion.ToString(), _userProfileDir, workloadSetVersion, useInstallStateOnly: true), offlineCache);
```

Comment:
This is probably related to the `useInstallStateOnly` parameter, but why is there an extra garbage collection here?  This probably needs an explanation in comments.

Reply:
It is! I'll add this in comment, but the quick idea is that we already checked if you asked for any manifests we didn't have installed and tried to install those, but we haven't checked if there were manifests you _didn't_ want that were installed. The way we agreed to do that was to just write the list of manifests we want to keep (i.e., those the customer asked for) to the install state file, then garbage collect. That does mean we need a garbage collection right at the end, however.